src/plugin/helpers/decorations.ts

```diff
@@ -0,0 +1,84 @@
+import type { Node } from "prosemirror-model";
+import { Mapping, StepMap } from "prosemirror-transform";
+import type { DecorationGroup, DecorationSource } from "prosemirror-view";
+import { DecorationSet } from "prosemirror-view";
+
+export const getMappedDecorationsFromSet = (
+  // A decoration set received, positioned relative to the outer editor. It may contain irrelevant decorations.
+  decorationSet: DecorationSet,
+  // The offset of the field from its containing element.
+  fieldOffsetFromElement: number,
+  // The node representing the current field in the ProseMirror document.
+  fieldNode: Node,
+  // The node representing the outer ProseMirror document containing the field.
+  document: Node
+) => {
+  // This field may receive decorations that will not apply to its range.
+  // Find the decorations in the context of the original document that should apply to this field, based on the position of
+  // the decorations in the original field.
+  // We must filter the decorations before we offset them via a 'map', otherwise there may be errors due to decorations
+  // being mapped outside of a valid range.
+  const relevantDecosFromOriginalDoc = decorationSet.find(
```

Comment:
Explicit type annotation `Decoration[]` might make the comment on line 25 more clear